File: 48-rotate_image.py
# time: O(n)
# space: O(1)
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        def transpose(m):
            for r in range(len(m)):
                for c in range(r, len(m[0])):
                    m[r][c], m[c][r] = m[c][r], m[r][c] 

        def mirror(m):
            for r in range(len(m)):
                for c in range(len(m[0]) // 2):
                    m[r][c], m[r][-c - 1] = m[r][-c - 1], m[r][c]


        transpose(matrix)
        mirror(matrix)

File: 48-rotate_image.cs
public class Solution {
    public void Transpose(int[][] m) {
        for (int r = 0; r < m.Length; r++) {
            for (int c = r; c < m.Length; c++) {
                (m[r][c], m[c][r]) = (m[c][r], m[r][c]);
            }
        }
    }

    public void Mirror(int[][] m) {
        for (int r = 0; r < m.Length; r++) {
            for (int c = 0; c < m.Length / 2; c++) {
                (m[r][c], m[r][m.Length - c - 1]) = (m[r][m.Length - c - 1], m[r][c]);
            }
        }
    }

    public void Rotate(int[][] m) {
        Transpose(m);
        Mirror(m);
    }
}
